src/main/java/org/commonjava/util/partyline/impl/infinispan/model/InfinispanJFS.java

```diff
@@ -0,0 +1,259 @@
+import com.sun.tools.javac.util.Context;
+import org.commonjava.util.partyline.callback.StreamCallbacks;
+import org.commonjava.util.partyline.lock.LockLevel;
+import org.commonjava.util.partyline.lock.local.LocalLockManager;
+import org.commonjava.util.partyline.lock.local.LocalLockOwner;
+import org.commonjava.util.partyline.lock.local.ReentrantOperation;
+import org.commonjava.util.partyline.lock.local.ReentrantOperationLock;
+import org.commonjava.util.partyline.spi.JoinableFile;
+import org.commonjava.util.partyline.spi.JoinableFilesystem;
+import org.infinispan.Cache;
+import org.infinispan.notifications.Listener;
+import org.infinispan.notifications.cachelistener.annotation.CacheEntryCreated;
+import org.infinispan.notifications.cachelistener.annotation.CacheEntryModified;
+import org.infinispan.notifications.cachelistener.event.CacheEntryEvent;
+
+import javax.transaction.HeuristicMixedException;
+import javax.transaction.HeuristicRollbackException;
+import javax.transaction.NotSupportedException;
+import javax.transaction.RollbackException;
+import javax.transaction.SystemException;
+import javax.transaction.TransactionManager;
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class InfinispanJFS
+                implements JoinableFilesystem
+{
+    private final LocalLockManager lockManager = new LocalLockManager();
+
+    private static final long WAIT_TIMEOUT = 100;
+
+    private String nodeKey;
+
+    private final Cache<String, FileMeta> metadataCache;
+
+    private final Cache<String, FileBlock> blockCache;
+
+    public InfinispanJFS( final String nodeKey, final Cache<String, FileMeta> metadataCache,
+                          final Cache<String, FileBlock> blockCache )
+    {
+        this.nodeKey = nodeKey;
+        this.metadataCache = metadataCache;
+        this.blockCache = blockCache;
+        this.nodeKey = nodeKey;
+    }
+
+    @Override
+    public JoinableFile getFile( final File file, final LocalLockOwner lockOwner, final StreamCallbacks callbacks,
+                                 final boolean doOutput, final ReentrantOperationLock opLock ) throws IOException
+    {
+        return new InfinispanJF( file, lockOwner, callbacks, doOutput, opLock, this );
+    }
+
+    @Override
+    public LocalLockManager getLocalLockManager()
+    {
+        return lockManager;
+    }
+
+    @Override
+    public void updateDominantLocks( String path, LockLevel level )
+                    throws SystemException, NotSupportedException, HeuristicRollbackException, HeuristicMixedException,
+                    RollbackException
+    {
+        TransactionManager transactionManager = metadataCache.getAdvancedCache().getTransactionManager();
+        transactionManager.begin();
+        FileMeta meta = metadataCache.get( path );
+        if ( level == null )
+        {
+            meta.removeLock( this.nodeKey );
+        }
+        else
```

Comment:
This looks right, except we should compare any pre-existing LockLevel for this node in the FileMeta and avoid doing a cache.put() if the level hasn't changed.